File: scripts/figma_scan.py
#!/usr/bin/env python3
import argparse
import json
import os
from urllib.request import Request, urlopen
from urllib.parse import urlencode
from datetime import datetime

VECTOR_TYPES = {
    'VECTOR',
    'BOOLEAN_OPERATION',
    'STAR',
    'POLYGON',
    'ELLIPSE',
    'RECTANGLE',
    'LINE',
}


def fetch_json(url: str, token: str):
    req = Request(url, headers={'X-Figma-Token': token})
    with urlopen(req) as res:
        data = res.read()
    return json.loads(data.decode('utf-8'))


def has_image_fill(node):
    fills = node.get('fills')
    if not isinstance(fills, list):
        return False
    return any(fill.get('type') == 'IMAGE' for fill in fills if isinstance(fill, dict))


def get_size(node):
    box = node.get('absoluteBoundingBox') or {}
    width = box.get('width')
    height = box.get('height')
    if width is None or height is None:
        return {'width': None, 'height': None}
    return {'width': int(round(width)), 'height': int(round(height))}


def traverse(node, parent_path, depth, nodes):
    name = node.get('name') or ''
    path_parts = parent_path + [name]
    node_id = node.get('id')

    child_has_image = False
    child_has_vector = False
    child_has_text = False

    for child in node.get('children') or []:
        summary = traverse(child, path_parts, depth + 1, nodes)
        child_has_image = child_has_image or summary['hasImageFill']
        child_has_vector = child_has_vector or summary['hasVector']
        child_has_text = child_has_text or summary['hasText']

    self_has_image = has_image_fill(node)
    self_has_vector = node.get('type') in VECTOR_TYPES
    self_has_text = node.get('type') == 'TEXT'

    summary = {
        'id': node_id,
        'name': name,
        'type': node.get('type'),
        'depth': depth,
        'visible': node.get('visible', True),
        'isMask': bool(node.get('isMask')),
        'hasImageFill': self_has_image or child_has_image,
        'hasVector': self_has_vector or child_has_vector,
        'hasText': self_has_text or child_has_text,
        'exportSettings': node.get('exportSettings') or [],
        'size': get_size(node),
        'path': ' / '.join(path_parts),
        'selfHasImage': self_has_image,
        'selfHasVector': self_has_vector,
        'selfHasText': self_has_text,
        'childrenCount': len(node.get('children') or []),
        'parentPath': ' / '.join(parent_path) if parent_path else None,
    }

    nodes.append(summary)
    return summary


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--token', default=os.getenv('FIGMA_TOKEN'))
    parser.add_argument('--file', default='Izgzufe5syG3KcDR7fdPVX')
    parser.add_argument('--node', default='314:8186')
    parser.add_argument('--output', default='data/figma-scan.json')
    args = parser.parse_args()

    if not args.token:
        raise SystemExit('Missing FIGMA_TOKEN. Provide --token or set FIGMA_TOKEN env var.')

    node_id = args.node.replace('-', ':')
    query = urlencode({'ids': node_id})
    url = f'https://api.figma.com/v1/files/{args.file}/nodes?{query}'
    data = fetch_json(url, args.token)
    root = data.get('nodes', {}).get(node_id, {}).get('document')
    if not root:
        raise SystemExit('Could not find node in response.')

    nodes = []
    traverse(root, [], 0, nodes)

    output_data = {
        'fileKey': args.file,
        'rootNodeId': node_id,
        'generatedAt': datetime.utcnow().isoformat() + 'Z',
        'nodeCount': len(nodes),
        'nodes': nodes,
    }

    os.makedirs(os.path.dirname(args.output), exist_ok=True)
    with open(args.output, 'w', encoding='utf-8') as f:
        json.dump(output_data, f, ensure_ascii=False, indent=2)

    print(f'Saved scan to {args.output} (nodes: {len(nodes)})')


if __name__ == '__main__':
    main()

File: scripts/figma-scan.mjs
#!/usr/bin/env node
import fs from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'

function getArgValue(flag) {
  const index = process.argv.indexOf(flag)
  if (index === -1) return null
  return process.argv[index + 1] || null
}

const token = getArgValue('--token') || process.env.FIGMA_TOKEN
const fileKey = getArgValue('--file') || 'Izgzufe5syG3KcDR7fdPVX'
const nodeId = (getArgValue('--node') || '314:8186').replace(/-/g, ':')
const output = getArgValue('--output') || path.resolve('data/figma-scan.json')

if (!token) {
  console.error('Missing FIGMA_TOKEN. Provide --token or set FIGMA_TOKEN env var.')
  process.exit(1)
}

async function fetchJson(url) {
  const res = await fetch(url, {
    headers: { 'X-Figma-Token': token },
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`Figma API error ${res.status}: ${body}`)
  }
  return res.json()
}

const url = new URL(`https://api.figma.com/v1/nodes/${fileKey}`)
url.searchParams.set('ids', nodeId)

const data = await fetchJson(url.toString())
const root = data?.nodes?.[nodeId]?.document
if (!root) {
  console.error('Could not find node in response.')
  process.exit(1)
}

const vectorTypes = new Set([
  'VECTOR',
  'BOOLEAN_OPERATION',
  'STAR',
  'POLYGON',
  'ELLIPSE',
  'RECTANGLE',
  'LINE',
])

function hasImageFill(node) {
  if (!Array.isArray(node.fills)) return false
  return node.fills.some((fill) => fill?.type === 'IMAGE')
}

function hasVectorSelf(node) {
  return vectorTypes.has(node.type)
}

function hasTextSelf(node) {
  return node.type === 'TEXT'
}

function getSize(node) {
  const box = node.absoluteBoundingBox
  if (!box) return { width: null, height: null }
  return { width: Math.round(box.width), height: Math.round(box.height) }
}

const nodes = []

function traverse(node, parentPath = [], depth = 0) {
  const visible = node.visible !== false
  const name = node.name || ''
  const pathParts = [...parentPath, name]
  const id = node.id

  let childHasImage = false
  let childHasVector = false
  let childHasText = false

  if (Array.isArray(node.children)) {
    for (const child of node.children) {
      const summary = traverse(child, pathParts, depth + 1)
      childHasImage = childHasImage || summary.hasImageFill
      childHasVector = childHasVector || summary.hasVector
      childHasText = childHasText || summary.hasText
    }
  }

  const selfHasImage = hasImageFill(node)
  const selfHasVector = hasVectorSelf(node)
  const selfHasText = hasTextSelf(node)

  const summary = {
    id,
    name,
    type: node.type,
    depth,
    visible,
    isMask: Boolean(node.isMask),
    hasImageFill: selfHasImage || childHasImage,
    hasVector: selfHasVector || childHasVector,
    hasText: selfHasText || childHasText,
    exportSettings: Array.isArray(node.exportSettings) ? node.exportSettings : [],
    size: getSize(node),
    path: pathParts.join(' / '),
    selfHasImage,
    selfHasVector,
    selfHasText,
    childrenCount: Array.isArray(node.children) ? node.children.length : 0,
  }

  nodes.push(summary)
  return summary
}

traverse(root, [], 0)

const outputData = {
  fileKey,
  rootNodeId: nodeId,
  generatedAt: new Date().toISOString(),
  nodeCount: nodes.length,
  nodes,
}

await fs.mkdir(path.dirname(output), { recursive: true })
await fs.writeFile(output, JSON.stringify(outputData, null, 2))

console.log(`Saved scan to ${output} (nodes: ${nodes.length})`)
